libs/SmartSync/src/com/salesforce/androidsdk/smartsync/util/SoqlSyncDownTarget.java

```diff
@@ -130,6 +138,80 @@ public JSONArray continueFetch(SyncManager syncManager) throws IOException, JSON
         return records;
     }
 
+    @Override
+    public Set<String> getListOfRemoteIds(SyncManager syncManager, Set<String> localIds) {
```

Comment:
We don't want to deal just with deleted records, we also want to purge out any records that no longer matches the SOQL query for the sync down.

So we should:
run the SOQL sync down with a modified SELECT to return only ids
Or
run the SOQL sync down with a modified SELECT to return only ids and an additional "and id in (local ids)" clause.
The second one is tricky because there are limits on SOQL query size.